File: Problem19_LongestPalindrome.java
// Time Complexity : O(n) -> n: length of the string
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    public int longestPalindrome(String s) {
        HashSet<Character> set = new HashSet<>();
        int count = 0;
        for(int i = 0; i < s.length(); i++) {
            if(set.contains(s.charAt(i))) {
                set.remove(s.charAt(i));
                count++;
            } else {
                set.add(s.charAt(i));
            }
        }
        
        if(!set.isEmpty()) 
            return count * 2 + 1;
        else
            return count * 2;
    }
}

File: Problem 19 Longest Palindrome.py
# Did this code successfully run on Leetcode: Yes
# Time Complexity: O(n)
# Space Complexity: O(1) since mapper size is constant with max size of 52 (small & capital letters)
# Logic: count the occurance of each character, only 1 odd character can come in palindrome others has to be even only.

class Solution:
    def longestPalindrome(self, s: str):
        mapper = {}
        count = 0
        isOdd = False
        for i in s:
            if i in mapper:
                mapper[i] += 1
            else:
                mapper[i] = 1
        for i in mapper:
            if not isOdd and i%2 == 1:
                isOdd = True
            count += (mapper[i]//2) * 2
        if isOdd:
            count += 1
        return count
